hw/ip/otbn/dv/rig/rig/gens/known_wdr.py

```diff
@@ -0,0 +1,133 @@
+                               'not the shape expected by the KnownWDR generator.')
+
+        self.wrd_op_type = self.bn_xor.operands[0].op_type
+        self.wrs_op_type = self.bn_xor.operands[1].op_type
+        self.imm_op_type = self.bn_xor.operands[4].op_type
+        assert self.imm_op_type.shift == 3
+
+        if not (isinstance(self.bn_not.operands[0].op_type, RegOperandType) and
+                self.bn_not.operands[0].op_type.reg_type == 'wdr' and
+                self.bn_not.operands[0].op_type.is_dest() and
+                isinstance(self.bn_not.operands[1].op_type, RegOperandType) and
+                self.bn_not.operands[1].op_type.reg_type == 'wdr' and
+                not self.bn_not.operands[1].op_type.is_dest() and
+                isinstance(self.bn_not.operands[2].op_type, EnumOperandType) and
+                isinstance(self.bn_not.operands[3].op_type, ImmOperandType)):
+            raise RuntimeError('BN.NOT instruction from instructions file is '
+                               'not the shape expected by KnownWDR generator. ')
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+        # Return None if this one of the last two instructions in the current gap
+        # because we need to either jump or do an ECALL to avoid getting stuck
+        # after executing both bn.xor and bn.not
+        if program.get_insn_space_at(model.pc) <= 3:
+            return None
+
+        # The bn.xor-bn.not pair takes 2 instructions
+        if program.space < 2:
+            return None
+
+        if model.fuel < 2:
+            return None
+
+        # Pick wrd any old way: we can write to any register. This should
+        # always succeed.
+        wrd_val_xor = model.pick_reg_operand_value(self.wrd_op_type)
+        if wrd_val_xor is None:
+                return None
+
+        # Picks a random operand value.
+        wrs_val_xor = model.pick_operand_value(self.wrs_op_type)
+        if wrs_val_xor is None:
+                return None
+
+        # Shift bits are hardcoded to 0 because wrs1=wrs2 has to happen
+        # for getting known value of zero as an answer.
+        shift_bits = self.imm_op_type.op_val_to_enc_val(0, model.pc)
+        assert shift_bits is not None
+
+        # Value of shift_type does not matter since shift_bits are hardcoded to 0
+        shift_type = model.pick_operand_value(self.bn_xor.operands[3].op_type)
+        assert shift_type is not None
+
+        # This value does not matter for this application
+        flg_group = model.pick_operand_value(self.bn_xor.operands[5].op_type)
+        assert flg_group is not None
+
+        # Result of this insn can be written to any register.
+        wrd_val_not = model.pick_reg_operand_value(self.wrd_op_type)
+        if wrd_val_not is None:
+                return None
+
+        # Trying to create a program where bn.xor and bn.not are consequently
+        # executed
```

Comment:
I'd probably drop this comment: it should be reasonably clear from the code. Also, you're not just trying: you're guaranteed to succeed at this point! :-)